code/modules/awaymissions/gateway.dm

```diff
@@ -327,3 +322,37 @@ GLOBAL_LIST_EMPTY(gateway_destinations)
 /obj/item/paper/fluff/gateway
 	info = "Congratulations,<br><br>Your station has been selected to carry out the Gateway Project.<br><br>The equipment will be shipped to you at the start of the next quarter.<br> You are to prepare a secure location to house the equipment as outlined in the attached documents.<br><br>--Nanotrasen Bluespace Research"
 	name = "Confidential Correspondence, Pg 1"
+
+/obj/effect/gateway_portal_effect
+	appearance_flags = KEEP_TOGETHER|TILE_BOUND|PIXEL_SCALE
+	mouse_opacity = MOUSE_OPACITY_TRANSPARENT
+	vis_flags = VIS_INHERIT_ID
+	layer = TABLE_LAYER //Slightly lower than gateway itself
+	var/alpha_icon = 'icons/obj/machines/gateway.dmi'
+	var/alpha_icon_state = "portal_mask"
+	var/datum/gateway_destination/our_destination
+
+
+/obj/effect/gateway_portal_effect/proc/setup_visuals(datum/gateway_destination/D)
+	our_destination = D
+	update_portal_filters()
+
+/obj/effect/gateway_portal_effect/proc/reset_visuals()
+	our_destination = null
+	update_portal_filters()
+
+/obj/effect/gateway_portal_effect/proc/update_portal_filters()
+	filters = null
+	vis_contents = null
+
+	if(!our_destination)
+		return
+
+	filters += filter(type="alpha", icon = icon(alpha_icon, alpha_icon_state), x = 32, y = 32)
```

Comment:
Could you keep track of filters and vis_contents instead of nulling them each time?
We have the `add_filter()` proc for this kind of thing, for example.

Reply:
ah yeah i forgot to add it

Reply:
we dont have such a system for vis_contents though and its definitely out of scope of this PR to add it to everything.